rust/protocol/src/sealed_sender.rs

```diff
@@ -879,8 +1031,133 @@ mod sealed_sender_v2 {
+    #[allow(clippy::unwrap_used)]
+    fn test_agreement_and_authentication() -> Result<()> {
+        // The sender and recipient each have a long-term identity key pair.
+        let sender_keypair = KeyPair::generate(&mut rand::thread_rng());
+        let sender_identity = IdentityKeyPair::try_from(sender_keypair).unwrap();
+
+        let recipient_keypair = KeyPair::generate(&mut rand::thread_rng());
+        let recipient_identity = IdentityKeyPair::try_from(recipient_keypair).unwrap();
+
+        // Generate random bytes used for our multi-recipient encoding scheme.
+        let m: [u8; MESSAGE_KEY_LEN] = (&mut rand::thread_rng()).gen();
+        // Derive an ephemeral key pair from those random bytes.
+        let ephemeral_keys = DerivedKeys::calculate(&m);
+        let ephemeral_public_key = ephemeral_keys.e.public_key;
+
+        // Encrypt the ephemeral key pair.
+        let sender_c_0 = apply_agreement_xor(
+            &ephemeral_keys.e,
+            &recipient_keypair.public_key,
+            Direction::Sending,
+            &m,
+        )?;
+        // Compute an authentication tag for the encrypted key pair.
+        let sender_at_0 = compute_authentication_tag(
+            &sender_identity,
+            recipient_identity.identity_key(),
+            Direction::Sending,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+
+        // The message recipient calculates the original random bytes and authenticates the result.
+        let recv_m = apply_agreement_xor(
+            &recipient_keypair,
+            &ephemeral_public_key,
+            Direction::Receiving,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_m.as_ref(), m.as_ref());
+
+        let recv_at_0 = compute_authentication_tag(
+            &recipient_identity,
+            sender_identity.identity_key(),
+            Direction::Receiving,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_at_0.as_ref(), sender_at_0.as_ref());
+
+        Ok(())
+    }
 }
 
+/// This method implements a single-key multi-recipient [KEM], defined in Manuel Barbosa's
+/// ["Randomness Reuse: Extensions and Improvements"].
+///
+/// [KEM]: https://en.wikipedia.org/wiki/Key_encapsulation
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// ### Contrast with Single-Recipient KEM
+/// The KEM scheme implemented by this method uses the "Generic Construction" in `4.1` of that
+/// paper, instantiated with [ElGamal encryption]. Unlike other message types sent over the wire,
+/// the encoded message returned by this method does *not* correspond to a specific protobuf schema
+/// due to the complex web of cryptographic data dependencies in different segments of the resulting
+/// encoded message. See the documentation of our [single-recipient KEM] for a discussion of the
+/// design tradeoffs of using this particular multi-recipient KEM.
```

Comment:
Eh. It could have been protobuf, but that's just wasting space in a format with so many fixed-width fields. We went back and forth but decided that the only thing protobuf would buy us is extensibility, and we could add another version instead.

I wrote up this document internally on the format; maybe some form of it belongs here or in the repository somewhere as well: 
[multi_recipient_sealed_sender.md](https://github.com/signalapp/libsignal-client/files/7255544/multi_recipient_sealed_sender.md).

Reply:
The rustdoc book mentions it does not have support for bare markdown files, which would certainly be convenient here. I think this looks brief and appropriate enough for the "Wire Format" section of the `sealed_sender_multi_recipient_encrypt()` docstring, so I'll try adapting it there.

Reply:
Hmmmm, rust-lang/rust#76709 certainly seems to imply it is possible to use rustdoc against markdown files directly. I will try to apply that feature.

Reply:
Ok, so it's possible to use *rustdoc* against markdown files directly as a command-line argument, but it seems that isn't integrated into `cargo doc` (either providing files as CLI arguments, or using raw markdown files). In 94d7ca0 I took the liberty of merging this doc into the "Wire Format" section of `sealed_sender_multi_recipient_encrypt()`, and linking into the resulting "The version byte" section from the "Wire Format" section of `sealed_sender_encrypt_from_usmc()`.

In these changes I have also referred to "Sealed Sender v1" and "Sealed Sender v2" instead of "the single-/multi-recipient KEM". See the "Wire Format" section for Sealed Sender v2 rendered at https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/fn.sealed_sender_multi_recipient_encrypt.html#wire-format.

Reply:
I wasn't necessarily thinking "in the repository" had to mean "included in the rustdoc", but it seems to work, especially since functions get their own pages. (Thanks for rendering the docs!) You *can* include a Markdown file using `#[doc = include_str!("foo.md")]` (if we're on a new enough Rust, h/t [the Rustdoc book](https://doc.rust-lang.org/rustdoc/the-doc-attribute.html)), but I think this is better anyway.